include/rmm/cuda_stream_view.hpp

```diff
@@ -0,0 +1,154 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <rmm/detail/error.hpp>
+
+#include <cuda_runtime_api.h>
+
+#include <atomic>
+#include <cstddef>
+#include <cstdint>
+
+namespace rmm {
+
+/**
+ * @brief Strongly-typed non-owning wrapper for CUDA streams with default constructor.
+ *
+ * This wrapper is simply a "view": it does not own the lifetime of the stream it wraps.
+ */
+class cuda_stream_view {
+ public:
+  constexpr cuda_stream_view()                        = default;
+  constexpr cuda_stream_view(cuda_stream_view const&) = default;
+  constexpr cuda_stream_view(cuda_stream_view&&)      = default;
+  constexpr cuda_stream_view& operator=(cuda_stream_view const&) = default;
+  constexpr cuda_stream_view& operator=(cuda_stream_view&&) = default;
+  ~cuda_stream_view()                                       = default;
+
+  // TODO disable construction from 0 after cuDF and others adopt cuda_stream_view
+  // cuda_stream_view(int)            = delete; //< Prevent cast from 0
+  // cuda_stream_view(std::nullptr_t) = delete; //< Prevent cast from nullptr
+  // TODO also disable implicit conversion from cudaStream_t
+
+  /**
+   * @brief Implicit conversion from cudaStream_t.
+   */
+  constexpr cuda_stream_view(cudaStream_t stream) noexcept : stream_{stream} {}
+
+  /**
+   * @brief Get the wrapped stream.
+   *
+   * @return cudaStream_t The wrapped stream.
+   */
+  constexpr cudaStream_t value() const noexcept { return stream_; }
+
+  /**
+   * @brief Explicit conversion to cudaStream_t.
+   */
+  explicit constexpr operator cudaStream_t() const noexcept { return value(); }
+
+  /**
+   * @brief Explicit conversion to uintptr_t.
+   */
+  explicit operator uintptr_t() const noexcept { return reinterpret_cast<uintptr_t>(value()); }
+
+  /**
+   * @brief Return true if the wrapped stream is the CUDA default stream (aka Null Stream).
+   */
+  constexpr bool is_default() const noexcept { return value() == cudaStreamDefault; }
+
+  /**
+   * @brief Return true if the wrapped stream is the CUDA per-thread default stream.
+   */
+  bool is_per_thread_default() const noexcept
+  {
+#ifdef CUDA_API_PER_THREAD_DEFAULT_STREAM
+    return value() == cudaStreamPerThread || value() == cudaStreamDefault;
+#else
+    return value() == cudaStreamPerThread;
+#endif
+  }
+
+  /**
+   * @brief Return true if the wrapped stream is explicitly the CUDA legacy default stream.
+   */
+  bool is_legacy_default() const noexcept { return value() == cudaStreamLegacy; }
```

Comment:
Should this similarly handle the situation where `value() == cudaStreamDefault` similar to what `is_per_thread_default()` does?